File: codes/list/面试题 02.07. 链表相交/intersection-of-two-linked-lists-lcci.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	curA := headA
	curB := headB
	var lenA, lenB int
	for ; curA != nil; curA = curA.Next {
		lenA++
	}
	for ; curB != nil; curB = curB.Next {
		lenB++
	}
	curA = headA
	curB = headB
	if lenA < lenB {
		curA, curB = curB, curA
		lenA, lenB = lenB, lenA
	}
	gap := lenA - lenB
	for ; gap > 0; gap-- {
		curA = curA.Next
	}
	for curA!=nil{
		if curA==curB{
			return curA
		}
		curA=curA.Next
		curB = curB.Next
	}
	return nil

}

File: codes/list/面试题 02.07. 链表相交/intersection-of-two-linked-lists-lcci.cpp
#include "ListNode.h"
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        auto curA = headA;
        auto curB = headB;
        int sizeA = 0, sizeB = 0;
        while (curA != nullptr) {
            sizeA++;
            curA = curA->next;
        }
        while (curB != nullptr) {
            sizeB++;
            curB = curB->next;
        }

        curA = headA;
        curB = headB;
        while (sizeA > sizeB) {
            curA = curA->next;
            sizeA--;
        }
        while (sizeB > sizeA) {
            curB = curB->next;
            sizeB--;
        }

        while(curA!=nullptr){
            if(curA==curB){
                return curA;
            }
            curA = curA->next;
            curB = curB->next;
        }
        return curA;
    }
};